Add post about porting software (#2)
managarm/managarm.github.io: _posts/2020-04-05-porting-software.markdown

([@avdgrinten](https://github.com/avdgrinten))
</span>

## Introduction
Porting existing software to new operating systems is not always an easy task. Yet, the general procedure of porting different software packages is often surprisingly similar. In this blog post, we walk through the process of porting the multi-system emulator [Mednafen](https://mednafen.github.io/) to the Managarm operating system. Managarm is a pragmatic microkernel-based OS with fully asynchronous I/O (see our [GitHub repository](https://github.com/managarm/managarm) for more information). While we discuss the specific example of Mednafen, this post can be seen as a general recipe for porting software to Managarm.

![Mednafen](/assets/2020-porting-software/mednafen-1.png)

**Overview.** Porting a package commonly consists of the following steps:
* Setting up the package's build system for (cross-)compilation to the new OS.
* Building the package and fixing any problems arising in the package's `./configure` script or during compilation.
* Implementing new OS (and libc) functionality to make the package work.

## Preparations and build system

> The first step to adding any package is locating the source code. This is usually not too hard. In the case of Mednafen, there is no public Git repository; hence, we will build from a source tarball. We decided to use the latest version at the time of writing, namely `1.24.0-UNSTABLE`. Now, the next step is to figure out how to compile the package. Mednafen uses the dated GNU Autotools build system; hence, compiling it amounts to:

Reviewer: `decided` -> `decide`

```
$ ./autogen.sh
$ ./configure --prefix=/usr
$ make
$ DESTDIR=... make install
```

> `autogen.sh` is a script to regenerate `configure` (via `autoreconf`); this step is not necessary on most modern build systems such as CMake or Meson. As usual, `configure` resolves dependencies and sets up the package's build directory while `make` and `make install` compile and install the package, respectively.

Reviewer: `on most modern` -> `for most modern` (?)

```yml
  - name: mednafen
    source:
      subdir: 'ports'
      url: 'https://mednafen.github.io/releases/files/mednafen-1.24.0-UNSTABLE.tar.xz'
      format: 'tar.xz'
      extract_path: 'mednafen'
      tools_required: [host-autoconf-v2.69, host-automake-v1.15,
                       host-libtool, host-pkg-config]
      regenerate:
        - args: ['./autogen.sh']
    tools_required: [system-gcc, host-pkg-config]
    pkgs_required: [sdl2]
    configure:
      - args:
        - '@THIS_SOURCE_DIR@/configure'
        - '--host=x86_64-managarm'
        - '--prefix=/usr'
        - '--with-sysroot=@SYSROOT_DIR@' # Set libtool's lt_sysroot.
        - '--without-libsndfile'
    build:
      - args: ['make', '-j@PARALLELISM@']
      - args: ['make', 'install']
        environ:
            DESTDIR: '@THIS_COLLECT_DIR@'
        quiet: true
```
As can be seen from this snippet, the only special options we need to pass to the `configure` script are:

>  -  `--disable-libsndfile`, as otherwise configuring fails due to libsndfile not being present,

Reviewer: ```suggestion
 -  `--disable-libsndfile`, as Managarm does not have a port of libsndfile yet.
```

Also, what about `--disable-debugger`?

>  - `--with-sysroot=@SYSROOT_DIR@`, to tell libtool about our system root (otherwise it will probably link against host libraries).

Reviewer: Maybe that's too much detail here (?) and we can remove this line (it already has a comment in the YAML snippet).

## Building the package

After adding the new package, the remaining steps are compiling and testing it, and fixing any problems that arise. Compiling packages for the first time consists of running the configure step (i.e `./configure`) to generate the build files, and running the build step to actually compile the sources.

### Configuration

> For Mednafen, the initial configure run fails with the following message:

Reviewer: ```suggestion
For Mednafen, the initial `configure` run fails with the following message:
```

configure:18177: error: *** pthreads not found!
```
This compilation failure is caused by the fact that [mlibc](https://github.com/managarm/mlibc) (the C library that Managarm uses) misses the corresponding function declarations. For Mednafen to work, this issue needs to be fixed.

**Fixing `sem_*`.** Adding `sem_*` functions to the C library requires (i) adding the headers and declarations, (ii) adding function definitions and (iii) actually implementing those functions. At this stage, we are primarily interested in making `configure` happy; hence, it is enough to add only stubs to mlibc (and not full implentations of `sem_*`). Looking at the [POSIX standard](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/semaphore.h.html), the required declarations look roughly like this:

```c
typedef struct __mlibc_sem_t {
	unsigned int __mlibc_count;
} sem_t;

int sem_init(sem_t *sem, int pshared, unsigned int initial_count);
int sem_destroy(sem_t *sem);
int sem_wait(sem_t *sem);
int sem_timedwait(sem_t *sem, const struct timespec *abstime);
int sem_post(sem_t *sem);
```
Note the `__mlibc` prefix for internal implementation details that are not specified by POSIX. 

> After adding these declarations to mlibc (i.e., to a new `semaphore.h` header), together with appropriate stubs, the configure step succeeded and the actual build could start.

Reviewer: `succeeded` -> `succeeds` and `could start` -> `can start`.

### Compiliation

> The build was going well with no errors up until the point when it hit some missing typedef. After quickly adding that to mlibc, the build continued, but failed again later on. This time it was missing the `iconv_*` functions. After initially thinking about porting `libiconv`, we realised the functions are a part of the standard library, and added stubs for them. These problems are fixed in the same way as the `sem_*` issues mentioned above, by adding new declarations and stubs to mlibc.¹ After adding them, we need to make the build system aware of new files, by adding them to one of the Meson files² and we need to recompile mlibc with `xbstrap install mlibc{,-headers} --rebuild`.

Reviewer: This paragraph should be changed to present tense.

 - implementing this function in the C library,
 - and handling threads in Managarm's POSIX.

To this end, a new request (`sys_clone`) to create a new thread is added to the POSIX server. In mlibc, the `pthread_create` function is modified to invoke the newly added `sys_clone` request. Before calling into the POSIX server, mlibc prepares the stack of the new thread and sets up the so-called *thread control block* (TCB) which stores all thread-local variables. Afterwards, control is handed over to the user provided function.

While we are still missing an implementation for `pthread_join` and possibly other thread related functions, it is not neccessary for just using the emulator, it only becomes a problem when you try to close it :^).

**POSIX semaphores.** Like pthread mutexes, POSIX semaphores are used to enforce mutual exclusion of concurrently executing threads. Since mlibc already supports the `pthread_mutex_*` functions, adding support for `sem_*` is comparatively simple. Indeed, both the pthread functions and POSIX semaphores are implemented on top of futexes.¹

> ¹ The implementation can be found in `options/posix/generic/semaphore-stubs.cpp`.

Reviewer: This should be ³.
